File: Homeworks/MatrixProduct/MatrixProduct.java
import java.util.*;
import java.math.*;
public class MatrixProduct {
    public static int MAX = 10010;
    public static int[][] A = new int[MAX][MAX];
    public static int[][] B = new int[MAX][MAX];
    public static int[][] ans = new int[MAX][MAX];
    
    public static class pdd{
        public double first,second;
        pdd(double first,double second){
            this.first=first;
            this.second=second;
        }    
    }
    
    public static void init(int n){
        for(int i = 0;i<n;i++){
            for(int j = 0;j<n;j++){
                A[i][j]=1;
                B[i][j]=2;
                ans[i][j]=0;
            }
        }
    }
    
    public static void showM(int M[][],int n){
        for(int i = 0;i<n;i++){
            for(int j = 0;j<n;j++){
                System.out.print(M[i][j]+" ");
            }
            System.out.println();
        }
    }
    public static void product(int n){
        for(int i = 0 ; i<n ; i++)
            for(int j = 0 ; j<n ; j++)
                for(int k = 0 ; k<n ; k++)
                    ans[i][j] = ans[i][j] + (A[i][k] * B[k][j]);
    }
    
    public static pdd prodTime(int n){
    double add=0,product=0;
    for(int i = 0 ; i<n ; i++)
        for(int j = 0 ; j<n ; j++)
            for(int k = 0 ; k<n ; k++){
                long start = System.nanoTime();
                int pd = (A[i][k] * B[k][j]);
                long end = System.nanoTime();
                product += (double)(end-start);
                start = System.nanoTime();
                ans[i][j] = ans[i][j] + pd;
                end = System.nanoTime();
                add += (double)(end-start);
            }
        return new pdd(add/(double)1e9,product/(double)1e9);
    }
    
    public static void main(String[] args) {   
        Scanner sc = new Scanner(System.in);
        System.out.println("Testing function - insert a value 0 <= n < 10010: ");
        int n = sc.nextInt();
        init(n);
        System.out.println("A:");showM(A,n);
        System.out.println("B:");showM(B,n);
        long start = System.nanoTime();
        product(n);
        long end = System.nanoTime();
        System.out.println("ans: ");showM(ans,n);
        System.out.println("time elapsed " + (end-start/((double)1e9) + " seconds."));
        
        System.out.println("----time performance analysis - each 10 steps----");
        System.out.println(" n \t avg(10) \t add_avg(10) \t prod_avg(10) ");       
        for (int i = 0 ; i <= 200;i+=10){
            init(i);//reset values of matrices
            double avg = 0, add_avg=0,prod_avg = 0;
            for(int j = 0;j<10;j++){
                pdd ot;
                start = System.nanoTime();
                ot = prodTime(i);
                end = System.nanoTime();
                avg += (end-start)/(double)1e9;
                   add_avg += ot.first;
                prod_avg += ot.second;
            }
            avg/=10.0;
            add_avg/=10.0;
            prod_avg/=10.0;
            System.out.printf("%d \t %.8f \t %.8f \t %.8f\n",i,avg,add_avg,prod_avg);
        }       
    }

}

File: Homeworks/MatrixProduct/MatrixProduct.py
import numpy as np
import time

#Product Function
def matrixProduct(A,B,n,m,k):
    C = np.zeros((n,k),dtype=int)
    for i in range(0,n):
        for j in range(0,k):
            for l in range(0,m):
                C[i][j] = C[i][j]+(A[i][l]*B[l][j])
    return C

#This is the same function but with some changes to measure time taken by operations(+,*).
def matrixProductTimes(A,B,n,m,k):
    C = np.zeros((n,k),dtype=int)
    add,multi = 0,0
    for i in range(0,n):
        for j in range(0,k):
            for l in range(0,m):
                to = time.clock()
                pr = A[i][l]*B[l][j]
                tf = time.clock()
                multi = multi + (tf-to)
                to = time.clock()
                C[i][j] = C[i][j]+pr
                tf = time.clock()
                add = add + (tf-to)
    return C,add,multi


print("Testing function - insert a value n: ")
hmm = (int)(input("> "))
# we define the matrix dimesion nxm for two matrix
n,m,k = hmm,hmm,hmm
#Matrix A,B
A = np.ones((n,m), dtype=np.int) #n*m
B = np.full((m,k), 2) #n*k
to= time.clock()
C = matrixProduct(A,B,n,m,k)
tf= time.clock()

print("A = \n",A)
print("B = \n",B)
print("AB = \n",C)
print ("time elapsed > ",tf-to," seconds.")

print ("---- PYTHON_3 time performance analysis - each 10 steps----")
print("n \t avg(10) \t add_avg(10) \t prod_avg(10)")
for i in range(0,200,10):
    avg,timeAdd,timeProd=0.0,0.0,0.0
    n=m=k=i
    A = np.ones((n,m), dtype=np.int) #n*m
    B = np.full((m,k), 2) #n*k
    for j in range(10):
        t0=time.clock()
        C,ad,pr = matrixProductTimes(A,B,n,m,k)
        tf=time.clock()
        avg += tf-t0
        timeAdd += ad
        timeProd += pr
    avg/=10.0
    timeAdd/=10.0
    timeProd/=10.0
    print(i,'\t',avg,'\t',timeAdd,'\t',timeProd)
